src/cargo/ops/common_for_install_and_uninstall.rs

```diff
@@ -1,45 +1,481 @@
+            .collect::<Vec<_>>();
+        for p in to_remove.iter() {
+            self.v1.remove(p);
+        }
+        // Add these bins.
+        let mut bins = bins.clone();
+        self.v1
+            .entry(pkg.package_id())
+            .and_modify(|set| set.append(&mut bins))
+            .or_insert(bins);
+    }
+
+    fn remove(&mut self, pkg_id: PackageId, bins: &BTreeSet<String>) {
+        let mut installed = match self.v1.entry(pkg_id) {
+            btree_map::Entry::Occupied(e) => e,
+            btree_map::Entry::Vacant(..) => panic!("v1 unexpected missing `{}`", pkg_id),
+        };
+
+        for bin in bins {
+            installed.get_mut().remove(bin);
+        }
+        if installed.get().is_empty() {
+            installed.remove();
+        }
+    }
+
+    fn save(&self, lock: &FileLock) -> CargoResult<()> {
+        let mut file = lock.file();
+        file.seek(SeekFrom::Start(0))?;
+        file.set_len(0)?;
+        let data = toml::to_string(self)?;
+        file.write_all(data.as_bytes())?;
+        Ok(())
     }
 }
 
+impl CrateListingV2 {
+    fn sync_v1(&mut self, v1: &CrateListingV1) -> CargoResult<()> {
+        // Make the `bins` entries the same.
+        for (pkg_id, bins) in &v1.v1 {
+            self.installs
+                .entry(*pkg_id)
+                .and_modify(|info| info.bins = bins.clone())
+                .or_insert_with(|| InstallInfo::from_v1(bins));
+        }
+        // Remove any packages that aren't present in v1.
+        let to_remove: Vec<_> = self
+            .installs
+            .keys()
+            .filter(|pkg_id| !v1.v1.contains_key(pkg_id))
+            .cloned()
+            .collect();
+        for pkg_id in to_remove {
+            self.installs.remove(&pkg_id);
+        }
+        Ok(())
+    }
+
+    fn package_for_bin(&self, bin_name: &str) -> Option<PackageId> {
+        self.installs
+            .iter()
+            .find(|(_, info)| info.bins.contains(bin_name))
+            .map(|(pkg_id, _)| *pkg_id)
+    }
+
+    fn mark_installed(
+        &mut self,
+        pkg: &Package,
+        bins: &BTreeSet<String>,
+        version_req: Option<String>,
+        opts: &CompileOptions<'_>,
+        target: String,
+        rustc: String,
+    ) {
+        // Remove bins from any other packages.
+        for info in &mut self.installs.values_mut() {
+            for bin in bins {
+                info.bins.remove(bin);
+            }
+        }
+        // Remove empty metadata lines. If only BTreeMap had `retain`.
+        let to_remove = self
+            .installs
+            .iter()
+            .filter_map(|(&p, info)| if info.bins.is_empty() { Some(p) } else { None })
+            .collect::<Vec<_>>();
+        for p in to_remove.iter() {
+            self.installs.remove(p);
+        }
+        // Add these bins.
+        if let Some(info) = self.installs.get_mut(&pkg.package_id()) {
+            for bin in bins {
+                info.bins.remove(bin);
```

Comment:
I may understand this later, but in a section called "add these bins" seems werid to see a `remove`

Reply:
Oops, that's the complete opposite of what it should be (should be append). I added some more stringent tests to catch this blunder.